### PR TITLE
Simplified data iterator. Removed source_length

### PR DESCRIPTION
**Pull request by fhieber**
source_length can be easily computed in the computation graph. This simplifies the data iterator quite a bit. 
Once we merged the transformer branch I will also remove the dependency of having source_length as an input argument to the inference encoder graph.

### CODE REVIEW

File: sockeye/data_io.py [hunk 1]
```diff
@@ -50,8 +50,8 @@ def define_buckets(max_seq_len: int, step=10) -> List[int]:
 
 def define_parallel_buckets(max_seq_len_source: int,
                             max_seq_len_target: int,
-                            bucket_width=10,
-                            length_ratio=1.0) -> List[Tuple[int, int]]:
+                            bucket_width: int = 10,
+                            length_ratio:float = 1.0) -> List[Tuple[int, int]]:
```
Comment (tdomhan) [inline sockeye/data_io.py]: nit-picky, but one line has a space between the colon and the type and the other doesn't.

File: sockeye/training.py [hunk 2]
```diff
@@ -94,7 +94,7 @@ def _build_module(self,
         Initializes model components, creates training symbol and module, and binds it.
         """
         source = mx.sym.Variable(C.SOURCE_NAME)
-        source_length = mx.sym.Variable(C.SOURCE_LENGTH_NAME)
+        source_length = mx.sym.sum(mx.sym.broadcast_not_equal(source, mx.sym.zeros((1,))), axis=1)
```
Comment (tdomhan) [inline sockeye/training.py]: how about adding an:
```
assert C.PAD_ID == 0
```
to catch issues if we ever changed the PAD_ID and to make it explicit that this is an assumption.

File: sockeye/data_io.py (hunk 1 above)
Comment (fhieber) [inline sockeye/data_io.py]: fixed

File: sockeye/training.py (hunk 2 above)
Comment (fhieber) [inline sockeye/training.py]: yes good point. We have that assertion somewhere in the data iterator as well

Comment (fhieber) [inline sockeye/training.py]: added

**Comment by fhieber** [comment]
no, I didn't see any performance differences locally. Its a fairly simple operation, compared to the unrolled RNNs its negligible I suppose.

**Pull request merged**